### PR TITLE
Further improve performance

### PR DESCRIPTION
**Pull request by tim-hoffman**

### CODE REVIEW

File: circuit_passes/src/bucket_interpreter/mod.rs [hunk 1]
```diff
@@ -245,23 +261,75 @@ impl<'a: 'd, 'd> BucketInterpreter<'a, 'd> {
...
+            env,
+            observe,
+            _compute_instruction,
+            _execute_instruction
+        );
+        error::add_loc_if_err(result, inst.as_ref())
     }
 
     /****************************************************************************************************
      * Private implemenation
```
Comment (0xddom) [inline circuit_passes/src/bucket_interpreter/mod.rs:315]: nit: typo

**Review by 0xddom — COMMENTED** [review]
I don't really understand what's going on, but I trust you that this improves performance. Because, if I understood it correctly, by splitting between compute and execute you avoid copying the env unnecessarily, correct? Do you have some numbers of the run time now?

Also, I was thinking. Do we still clone the whole IR every time we run a pass? Maybe we could improve performance further if we move the IR and only create new copies of things if we actually modify them. I don't know if it will work since we need the IDs for each bucket but may be worth a try

**Comment by tim-hoffman** [comment]
You got it. Profiling showed that Env clones and destructors were taking the most time by far.
I've replaced "execute" with "compute" in many cases where the bucket cannot update the Env. The `compute_or_execute` macro helps in cases where the Env _might_ be updated but we don't end up using the updated Env so it's safe to try the "compute" approach first but if an Env modifying instruction is encountered, fall back to the "execute" approach.

**Comment by tim-hoffman** [comment]
Each pass does still reconstruct the IR via the `transform` methods but I haven't seen that as a bottleneck in profiling so far.

**Comment by 0xddom** [comment]
> Each pass does still reconstruct the IR via the `transform` methods but I haven't seen that as a bottleneck in profiling so far.

It's not only performance but memory usage, but we can leave it for future work then, if it's not as pressing. This PR has enough stuff already

**Review by 0xddom — APPROVED** [review]

**Pull request merged**